jwst/resample/resample.py

```diff
@@ -1,357 +1,585 @@
-
-            iscale = self._get_intensity_scale(img)
-            log.debug(f'Using intensity scale iscale={iscale}')
-
-            inwht = resample_utils.build_driz_weight(
-                img,
-                weight_type=self.weight_type,
-                good_bits=self.good_bits
+        if self.blendheaders:
+            self._blender = ModelBlender(
+                blend_ignore_attrs=[
+                    'meta.photometry.pixelarea_steradians',
+                    'meta.photometry.pixelarea_arcsecsq',
+                    'meta.filename',
+                ]
             )
+        self.output_jwst_model = None
 
-            # apply sky subtraction
-            blevel = img.meta.background.level
-            if not img.meta.background.subtracted and blevel is not None:
-                data = img.data - blevel
-            else:
-                data = img.data
+    def resample_group(self, indices):
+        """ Resample multiple input images that belong to a single
+        ``group_id`` as specified by ``indices``.
 
-            xmin, xmax, ymin, ymax = resample_utils._resample_range(
-                data.shape,
-                img.meta.wcs.bounding_box
-            )
-            pixmap = resample_utils.calc_gwcs_pixmap(
-                img.meta.wcs,
-                self.output_wcs,
-                img.data.shape,
-            )
+        Parameters
+        ----------
+        indices : list
 
-            driz.add_image(
-                data=data,
-                exptime=img.meta.exposure.exposure_time,  # GWCSDrizzle.add_image param default was 1.0
-                pixmap=pixmap,
-                scale=iscale,
-                weight_map=inwht,
-                wht_scale=1.0,  # hard-coded for JWST count-rate data
-                pixfrac=self.pixfrac,
-                in_units="cps",  # GWCSDrizzle.add_image param default
-                xmin=xmin,
-                xmax=xmax,
-                ymin=ymin,
-                ymax=ymax,
-            )
-            del data
-
-            # make an approximate error image by drizzling it
-            # in the same way the image is handled
-            if compute_error:
-                driz_error.add_image(
-                    data=img.err,
-                    exptime=img.meta.exposure.exposure_time,  # GWCSDrizzle.add_image param default
-                    pixmap=pixmap,
-                    scale=iscale,
-                    weight_map=inwht,
-                    wht_scale=1.0,  # hard-coded for JWST count-rate data
-                    pixfrac=self.pixfrac,
-                    in_units="cps",  # GWCSDrizzle.add_image param default
-                    xmin=xmin,
-                    xmax=xmax,
-                    ymin=ymin,
-                    ymax=ymax,
-                )
-            input_models.shelve(img, index, modify=False)
-            del img
+        """
+        if self.output_jwst_model is not None:
+            self.reset_arrays(reset_output=True)
```

Comment:
```suggestion
        if self.finalized():
            self.reset_arrays(reset_output=True, n_output_models=len(indicies))
```
Is this checking if the instance of `ResampleImage` was previously `finalized`? If so, would this be easier to read?

Reply:
`finalize()` sets `self._finalized` flag only when it is called with `free_memory=True` which frees intermediate arrays used , as an example, for variance computations. Once those are deleted, calling `add_model()` should not be allowed. That is what `finalized()` is checking. `self.output_jwst_model is not None` on the other hand will reset arrays regardless of the status of intermediate arrays.

And anyway, what is the issue with `self.output_jwst_model is not None`?

Reply:
Thanks for the details.

I'm trying to understand this to adapt it to romancal (where at the moment there is no `output_jwst_model` and I'd like to avoid pre-allocating an output model since stcal pre-allocates the data arrays).

Is it a bug in the stcal PR that `self._finalized` is set to `True` for every call to `finalize`?
https://github.com/mcara/stcal/blob/c5c1a74a90830295104ec9568ef784e4d32e5b29/src/stcal/resample/resample.py#L1123
is it also a bug that the jwst subclass always overrides `free_memory` to True:
https://github.com/mcara/jwst/blob/2c53cccaba3b4de756c01c665c4c8f6ec0a074a7/jwst/resample/resample.py#L484

If `finalized` can't be trusted here than I think the documentation in the stcal PR could use some improvements. I read the `reset_arrays` docstring:
https://github.com/mcara/stcal/blob/c5c1a74a90830295104ec9568ef784e4d32e5b29/src/stcal/resample/resample.py#L831
to mean that `reset_arrays` needs to be called if an instance was already `finalized`. Since `resample_group` here calls `finalize` I took the documentation to mean that a second call (but not the first) of `resample_group` will need to call `reset_arrays` (because the first call called `finalize`). If that's not the case than the documentation looks incorrect.

Reply:
Good points. Let me investigate this.

Reply:
Fixed this in a recent commit. Basically introduced a new "state" - locked to indicate that intermediate arrays are gone and adding new images is not possible. `finalize()` in itself does not lock the object (unless it is called with a setting that may free memory).